\documentclass{report}\usepackage[]{graphicx}\usepackage[]{color}
%% maxwidth is the original width if it is less than linewidth
%% otherwise use linewidth (to make sure the graphics do not exceed the margin)
\makeatletter
\def\maxwidth{ %
  \ifdim\Gin@nat@width>\linewidth
    \linewidth
  \else
    \Gin@nat@width
  \fi
}
\makeatother

\definecolor{fgcolor}{rgb}{0.345, 0.345, 0.345}
\newcommand{\hlnum}[1]{\textcolor[rgb]{0.686,0.059,0.569}{#1}}%
\newcommand{\hlstr}[1]{\textcolor[rgb]{0.192,0.494,0.8}{#1}}%
\newcommand{\hlcom}[1]{\textcolor[rgb]{0.678,0.584,0.686}{\textit{#1}}}%
\newcommand{\hlopt}[1]{\textcolor[rgb]{0,0,0}{#1}}%
\newcommand{\hlstd}[1]{\textcolor[rgb]{0.345,0.345,0.345}{#1}}%
\newcommand{\hlkwa}[1]{\textcolor[rgb]{0.161,0.373,0.58}{\textbf{#1}}}%
\newcommand{\hlkwb}[1]{\textcolor[rgb]{0.69,0.353,0.396}{#1}}%
\newcommand{\hlkwc}[1]{\textcolor[rgb]{0.333,0.667,0.333}{#1}}%
\newcommand{\hlkwd}[1]{\textcolor[rgb]{0.737,0.353,0.396}{\textbf{#1}}}%
\let\hlipl\hlkwb

\usepackage{framed}
\makeatletter
\newenvironment{kframe}{%
 \def\at@end@of@kframe{}%
 \ifinner\ifhmode%
  \def\at@end@of@kframe{\end{minipage}}%
  \begin{minipage}{\columnwidth}%
 \fi\fi%
 \def\FrameCommand##1{\hskip\@totalleftmargin \hskip-\fboxsep
 \colorbox{shadecolor}{##1}\hskip-\fboxsep
     % There is no \\@totalrightmargin, so:
     \hskip-\linewidth \hskip-\@totalleftmargin \hskip\columnwidth}%
 \MakeFramed {\advance\hsize-\width
   \@totalleftmargin\z@ \linewidth\hsize
   \@setminipage}}%
 {\par\unskip\endMakeFramed%
 \at@end@of@kframe}
\makeatother

\definecolor{shadecolor}{rgb}{.97, .97, .97}
\definecolor{messagecolor}{rgb}{0, 0, 0}
\definecolor{warningcolor}{rgb}{1, 0, 1}
\definecolor{errorcolor}{rgb}{1, 0, 0}
\newenvironment{knitrout}{}{} % an empty environment to be redefined in TeX

\usepackage{alltt}
\usepackage[a4paper]{geometry} 
\geometry{a4paper, 
tmargin=2cm, 
bmargin=2cm, 
lmargin=1cm, 
rmargin=1cm, 
headheight=1.5em, 
headsep=0.5cm, 
footskip=1cm} 
\usepackage[ddmmyyyy,12hr]{datetime}

\usepackage{fancyhdr}
\pagestyle{fancy}
\lhead{
\textbf{\LARGE{BRCA2:c.8107A\textgreater G (p.Thr2703Ala)}}
}
\rhead{\textbf{\Large{CaVaDa variant report}}}

\usepackage{lastpage}
\cfoot{Page $\thepage of \pageref{LastPage}$}
\usepackage[normalem]{ulem}
\usepackage{hyperref}
\usepackage{longtable}
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\begin{document}
\begin{table}[h!]
\begin{tabular}{rp{12cm}rl}
\textit{Automated class}: & 2-Likely non-pathogenic &  \textit{Date:} & \today \\
\textit{Justification}: & missense &  \textit{Time:} & \currenttime \\
\textit{Curated class}: &  & \\
\end{tabular}
\end{table}

\smallskip



\smallskip

% latex table generated in R 3.3.2 by xtable 1.8-2 package
% Tue Jan  3 12:52:54 2017
\begin{longtable}{|rp{6cm}|}
   \hline
\textit{Gene:} & BRCA2 \\ 
  \textit{Ensembl transcript:} & ENST00000380152 \\ 
  \textit{RefSeq transcript:} & NM\_000059.3 \\ 
  \textit{Chromosome:} & 13 \\ 
  \textit{Genomic position:} & 32937446 \\ 
  \textit{Reference allele:} & A \\ 
  \textit{Variant allele:} & G \\ 
  \textit{HGVS cDNA:} & c.8107A$>$G \\ 
  \textit{HGVS protein:} & p.Thr2703Ala \\ 
  \textit{HGVS protein (1 letter):} & p.T2703A \\ 
  \textit{Variant location:} & exon \\ 
  \textit{Exon/intron number:} & exon 18 of 27 \\ 
  \textit{Variant type:} & substitution \\ 
  \textit{Variant coding effect:} & missense \\ 
  \textit{Most distant conserved species:} & Zebrafish \\ 
  \textit{AlignGVGD class:} & C0 \\ 
  \textit{AlignGVGD: variation (GV):} & 57.75 \\ 
  \textit{AlignGVGD: deviation (GD):} & 49.44 \\ 
  \textit{SIFT prediction:} & Deleterious \\ 
  \textit{MAPP prediction:} & bad \\ 
  \textit{CADD scaled score [0-99]:} & 8.757 \\ 
  \textit{SuSPect score [0-100]:} & 31 \\ 
  \textit{MutationTaster prediction:} & polymorphism \\ 
  \textit{Polyphen2 HumVar prediction:} & benign \\ 
  \textit{Distance to nearest splice site:} & 131 \\ 
  \textit{Nearest splice site:} & 3' \\ 
  \textit{Wild Type MaxEnt Score:} & 11.4993 \\ 
  \textit{Variant MaxEnt Score:} & 11.4993 \\ 
  \textit{\% change MaxEnt Score:} & 0 \\ 
  \textit{Wild Type NNS Score:} & 0.980273 \\ 
  \textit{Variant NNS Score:} & 0.980273 \\ 
  \textit{\% change NNS Score:} & 0 \\ 
  \hline
\end{longtable}


\smallskip

\textsf{}%
\framebox{\begin{minipage}[t]{0.9\columnwidth}%
\textsf{\emph{This resource is currently under development
and is for research use only.}}

\medskip

\textsf{\textbf{\scriptsize{CLASSIFICATION:}}}\textsf{\scriptsize{
1-Non-pathogenic, 2-Likely non-pathogenic, 3-Uncertain/review, 4-Likely
pathogenic, 5-Pathogenic}}
{\scriptsize \par}

\textsf{\scriptsize{The automated class is generated by the gene-specific
decision-tree, with an associated classification\_justification.}}
{\scriptsize \par}

\textsf{\scriptsize{The curated class is assigned following clinician review +/- literature search.}}
{\scriptsize \par}

\textsf{\scriptsize{Classifications, both automated and curated, are derived from integration of the variant level data according to objective criteria.}}
{\scriptsize \par}

\medskip

\textsf{\scriptsize{The variant classification system includes and the variant report may make reference to the following sources of information: }}
{\scriptsize \par}

\smallskip

{\raggedright{}
\textsf{\textbf{\textsl{\scriptsize{ALL GENES:}}}}
\textsf{\scriptsize{Exome Sequencing Project (\href{http://evs.gs.washington.edu/EVS/}{http://evs.gs.washington.edu/EVS/}), 1000 Genomes (\href{http://www.1000genomes.org/}{http://www.1000genomes.org/}), SIFT, MAPP, }}
\textsf{\tiny{\textsuperscript{{\tiny{1}}}}}\textsf{\scriptsize{AlignGvGD, SpliceSiteFinder, MaxEnt, NNSplice (\href{http://www.interactive-biosoftware.com/software/alamut-ht/overview}{http://www.interactive-biosoftware.com/software/alamut-ht/overview}), MutationTaster (\href{http://www.mutationtaster.org/}{http://www.mutationtaster.org/}), SuSPect (\href{http://www.sbg.bio.ic.ac.uk/suspect/about.html}{http://www.sbg.bio.ic.ac.uk/suspect/about.html}),}}
\textsf{\tiny{\textsuperscript{{\tiny{2}}}}}\textsf{\scriptsize{CADD (\href{http://cadd.gs.washington.edu/}{http://cadd.gs.washington.edu/}), HGMD (Human Gene Mutation Database \href{http://www.biobase-international.com/product/hgmd}{http://www.biobase-international.com/product/hgmd}), DMuDB (\href{http://www.ngrl.org.uk/Manchester/projects/dmudb}{http://www.ngrl.org.uk/Manchester/projects/dmudb})}}
}
{\scriptsize \par}

\smallskip

{\raggedright{}
\textsf{\textbf{\textsl{\scriptsize{GENE-SPECIFIC RESOURCES:}}}}
\textsf{\scriptsize{LOVDs (Leiden Open Variation Database \href{http://chromium.liacs.nl/}{http://chromium.liacs.nl/}), BIC (Breast Cancer Information Core \href{http://research.nhgri.nih.gov/bic/}{http://research.nhgri.nih.gov/bic/}),}}
\textsf{\textbf{\uline{\scriptsize{Multifactorial analyses:}}}}
\textsf{\scriptsize{
\textsuperscript{{\tiny{3}}}\href{http://www.ncbi.nlm.nih.gov/pubmed/17924331}{Easton, D.F. et al. Am J Hum Genet 81,873-83 (2007)},
\textsuperscript{{\tiny{4}}}\href{http://www.ncbi.nlm.nih.gov/pubmed/21990134}{Lindor, N.M. et al. Hum Mutat 33, 8-21 (2012)},
\href{http://www.ncbi.nlm.nih.gov/pubmed/21990165}{IARC} (\href{http://brca.iarc.fr/LOVD/}{http://brca.iarc.fr/LOVD/}), InSiGHT (\href{http://www.ncbi.nlm.nih.gov/pubmed/24362816}{Thompson, B et al. Nat Genetics 46, 107-114 (2014)},}}
\textsf{\textbf{\uline{\scriptsize{Functional analyses:}}}}
\textsf{\scriptsize{
\href{http://www.ncbi.nlm.nih.gov/pubmed/23108138}{Guidugli, L. et al. Cancer Res 73, 265-75 (2013)},
\href{http://www.ncbi.nlm.nih.gov/pubmed/23867111}{Bouwman,P Cancer Discov 3(10):1142-5 (2013)},
\href{http://www.ncbi.nlm.nih.gov/pubmed/20020535}{Drost, M et al Huma Mutat 31(3):247-53 (2010)},
\href{http://www.ncbi.nlm.nih.gov/pubmed/22102614}{Drost, M et al Hum Mutat 33(3):488-94 (2012)},}}
\textsf{\textbf{\uline{\scriptsize{Splicing analyses:}}}}\textsf{\scriptsize{}}
\textsf{\tiny{\textsuperscript{{\tiny{5}}}}}\textsf{\scriptsize{\href{http://www.ncbi.nlm.nih.gov/pubmed/22505045}{Houdayer, C. et al. Hum Mutat 33, 1228-38 (2012)}{[}1S, 2S, 3S{]}, \href{http://www.ncbi.nlm.nih.gov/pubmed/23893897}{Walker LC et al. Hum Mutat 34 (10):1424-31 (2013)}}}
}
{\scriptsize \par}

\medskip

\textsf{\tiny{\textsuperscript{{\tiny{1}}}}}
\textsf{\scriptsize{C0 = least conserved ... C65= most conserved }}
{\scriptsize \par}

\textsf{\tiny{\textsuperscript{{\tiny{2}}}}}
\textsf{\scriptsize{10 is assigned to variants belonging in the top 10\% of deleteriousness; 20 is assigned to variants belonging in the top 1\%; 30 is assigned to variants belonging in the top 0.1\%; ... 90 is assigned to variants belonging in the top 0.0000001\% }}
{\scriptsize \par}

\noindent 
\textsf{\tiny{\textsuperscript{{\tiny{3}}}}}
\textsf{\scriptsize{Odds of causality$>$40:1 = causal, Odds of neutrality$>$100:1 = neutral }}
{\scriptsize \par}

\textsf{\tiny{\textsuperscript{{\tiny{4}}}}}
\textsf{\scriptsize{Posterior probability of pathogenicity$>$0.99 = pathogenic, $>$0.95 = likely pathogenic, $<$0.05 = likely non pathogenic, $<$0.01 = non pathogenic }}
{\scriptsize \par}

\textsf{\tiny{\textsuperscript{{\tiny{5}}}}}
\textsf{\scriptsize{Class 1S: no effect on splicing; Class 2S: (1) impact on alternative splicing and (2) leaky splice site mutation, partial effect; Class 3S: severe impact on splicing: (1) the mutant allele does not produce the wild-type transcript, complete effect and (2) AG/GT disruption.}}%
{\scriptsize \par}
\end{minipage}}

\end{document}
